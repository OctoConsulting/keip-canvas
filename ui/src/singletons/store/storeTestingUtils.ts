import { renderHook } from "@testing-library/react"

// Zustand mock import (see '/home/aasaad/projects/keip-canvas/ui/__mocks__/zustand.ts')
// @ts-expect-error the reset function does not exist as a zustand export.
// It is added to the mock as a convenience for setup and cleanup of the store in the tests
import { resetStore } from "zustand"

// WARNING: In test files, import the reset method last, to ensure it's properly initialized.
export const resetMockStore = resetStore as (s: object) => void

export function renderAndUnwrapHook<T>(hookFn: () => T): T {
  const { result } = renderHook(hookFn)
  return result.current
}
