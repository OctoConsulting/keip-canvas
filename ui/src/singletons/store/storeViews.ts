import { Edge } from "reactflow"
import { EipFlowNode, Layout } from "../../api/flow"
import { EipId } from "../../api/generated/eipFlow"
import { useAppStore } from "./appStore"

// Warning: the following exports provide non-reactive access to the store's state
export const getNodesView = (): readonly EipFlowNode[] =>
  useAppStore.getState().nodes

export const getEdgesView = (): readonly Edge[] => useAppStore.getState().edges

export const getLayoutView = (): Readonly<Layout> =>
  useAppStore.getState().layout

export const getEipId = (nodeId: string): Readonly<EipId> | undefined =>
  useAppStore.getState().eipConfigs[nodeId]?.eipId
