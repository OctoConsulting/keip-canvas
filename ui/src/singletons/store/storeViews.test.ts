import { act } from "@testing-library/react"
import { beforeEach, expect, test, vi } from "vitest"
import { childrenBreadthTraversal, childrenDepthTraversal } from "./storeViews"
import nestedChildFlow from "./testdata/store-initializers/nestedChildFlow.json"

import { resetMockStore } from "./storeTestingUtils"

vi.mock("zustand")

const ROOT_ID = "FL5Tssm8tV"

beforeEach(() => {
  act(() => {
    resetMockStore(nestedChildFlow)
  })
})

test("traverse children depth-first search", () => {
  const nodes = [...childrenDepthTraversal(ROOT_ID)]
  expect(nodes).toMatchSnapshot()
})

test("traverse children breadth-first search", () => {
  const nodes = [...childrenBreadthTraversal(ROOT_ID)]
  expect(nodes).toMatchSnapshot()
})
