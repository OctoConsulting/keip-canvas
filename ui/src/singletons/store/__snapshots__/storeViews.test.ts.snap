// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`traverse children breadth-first search 1`] = `
[
  {
    "id": "FL5Tssm8tV",
    "parentId": null,
  },
  {
    "id": "43FAdk5SdBR",
    "parentId": "FL5Tssm8tV",
  },
  {
    "id": "hSZG9DyYy8q",
    "parentId": "FL5Tssm8tV",
  },
  {
    "id": "DD7ZbYQ3AST",
    "parentId": "FL5Tssm8tV",
  },
  {
    "id": "9OymVfY4n4p",
    "parentId": "43FAdk5SdBR",
  },
  {
    "id": "5x4O_wYO-tM",
    "parentId": "DD7ZbYQ3AST",
  },
  {
    "id": "LYutVnb04dY",
    "parentId": "DD7ZbYQ3AST",
  },
  {
    "id": "gCL0YkrFGgW",
    "parentId": "9OymVfY4n4p",
  },
]
`;

exports[`traverse children depth-first search 1`] = `
[
  {
    "id": "FL5Tssm8tV",
    "parentId": null,
  },
  {
    "id": "DD7ZbYQ3AST",
    "parentId": "FL5Tssm8tV",
  },
  {
    "id": "LYutVnb04dY",
    "parentId": "DD7ZbYQ3AST",
  },
  {
    "id": "5x4O_wYO-tM",
    "parentId": "DD7ZbYQ3AST",
  },
  {
    "id": "hSZG9DyYy8q",
    "parentId": "FL5Tssm8tV",
  },
  {
    "id": "43FAdk5SdBR",
    "parentId": "FL5Tssm8tV",
  },
  {
    "id": "9OymVfY4n4p",
    "parentId": "43FAdk5SdBR",
  },
  {
    "id": "gCL0YkrFGgW",
    "parentId": "9OymVfY4n4p",
  },
]
`;
